### ISSUE TITLE
Enhancement: Single Webserver Instance per test class

### ISSUE THREAD

**Issue by Pytry** [spring-projects/spring-boot#8827]
When I define a set of integration test in a single test class annotated with "@SpringBootTest", "@RunWith(SpringRunner.class)", and "@WebAppConfiguration", a new instance of tomcat is started for each "@Test" annotated method.

It would be great if there were a way to indicate that only one application server instance should be created to run all tests in the class for unit tests (which ever is configured for that module or project). This would really speed up our builds and integration testing. It would still be useful to be able to start a new instance per method as well, and I think it should remain the default behavior.

Currently, I can get around this by having a single method call all the other tests, but this does not give us the granular control available using the @Test annotations (timeouts, expected exceptions etc).

**Comment by wilkinsona** [comment]
Caching of the application context means that this is what should already happen by default. Can you provide a small sample that illustrates the behaviour that you have described?

**Comment by spring-issuemaster** [comment]
If you would like us to look at this issue, please provide the requested information. If the information is not provided within the next 7 days this issue will be closed.

**Comment by spring-issuemaster** [comment]
Closing due to lack of requested feedback. If you would like us to look at this issue, please provide the requested information and we will re-open the issue.

**Comment by Pytry** [comment]
Sorry for taking so long. Obviously, I was not aware that the context caching was the default behavior since I had seen it starting a new context (as described in the initial ticket), or I may have logged a bug report instead of an enhancement request.

I had an example in progress that ~~was~~ starting a new context for each "\@Test" annotation, but when I added a logback file only appender to it with the intention of saving the console output to upload for you to see, suddenly the caching was working properly. I've been spending some time trying to get it back to that state, but I can't figure out what I did (reverting the logging configuration changes did not work).

Could someone tell me more about how Spring Boot caches the context, or point me to the right classes or package to peruse?

OS: Windows Server 2012 R2
IDE: IntelliJ 2017.1
Java: 1.8.0_121
Maven: 3.3.6 (pre-requisite of 3.3.3)
p.s. Maybe I'm just hallucinating?

**Comment by Pytry** [comment]
I figured it out; I was just hallucinating.

https://github.com/Pytry/spring-boot-8827

The problem was that the context could not load because I had not finished the example completely (I was missing an @Bean annotation) which would of course cause the test runner to attempt to load the context when running each test method. I probably saw this before whenever I was building a new application and then just never paid attention after all the errors were fixed. 

I would still like to dig into the context caching to see if it would be possible to cache the context across test classes, perhaps by setting up a global context?

For example, perhaps an annotation (or a new property on @SpringBootTest) like this:

    @GlobalWebContext("restDocumentation")

or

    @SpringBootTest( globalWebContext = "restDocumentation")

Any class that had the same value assigned would use the same context. The default value would be the name of the class. So in the following examples, all the methods in the first two classes would share a test context, but only the one method in the third class would share that same context.
    
    @SpringBootTest
    public class TestSuiteOne {
        //All test methods run in "testSuiteOne" context.
    }

    @SpringBootTest( globalWebContext = "testSuiteOne")
    public class TestSuiteTwo {
        //All test methods run in "testSuiteOne" context.
    }

    @SpringBootTest
    public class TestSuiteThree {

        @Test
        public void contextNotShared(){}

        @Test
        @GlobalWebContext("testSuiteOne")        
        public void contextShared(){}
    }

Anything to make CI run faster (if indeed this does make it faster).
